### PR TITLE
fix: set omp affinity can't be called multiple times.

### PR DESCRIPTION
**Pull request by i8run**
This patch fix the `setOmpAffinity` issue, which can't be called multiple times previously. And 2nd, add a `stats()` to get the current affinity settings of all java thread.

Affinity class usage,

1. `Affinity.setAffinity()` in java thread. It will select a core who has minimal thread bound to bind.
2. `Affinity.setAffinity(arrayOfCores)` in java thread. It will force bind current java thread to cores. In cgroup (docker) environment, this available cores are limited by the cgroup settings.  And if user set invalid core id, it will throw a `RuntimeException`.
3. `Affinity.setOmpAffinity()`. It will set the current thread and omp thread fork from current thread to relative cores one by one.
3. `Affinity.stats()`. It will return a map of current affinity settings. The key of map is the core id.
4. `Affinity.getAffinity()` will return the affinity settings of current thread. `Affinity.getOmpAffinity()` will return affinity settings of current thread and the relative omp threads.

### CODE REVIEW

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/Affinity.java [hunk 1]
```diff
@@ -14,8 +17,8 @@
         IMPL = new LinuxAffinity(CpuInfo.INSTANCE.getLogicalProcessorCount());
     }
 
-    public static int setAffinity() {
-        return IMPL.setAffinity();
+    public static void setAffinity() {
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/Affinity.java]: why not returning the result ? how we can know if the binding works or not ?

Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/Affinity.java]: If not, the result will throw an exception.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 2]
```diff
@@ -2,84 +2,185 @@
...
-        for (int i = 0; i < availableCPUs.length; i++) {
-            if (!availableCPUs[i]) {
-                ret = i;
-                int[] set = {i};
-                setAffinity0(set);
-                availableCPUs[i] = true;
-                break;
+    public synchronized void setAffinity() {
+        int min = Integer.MAX_VALUE;
+        int theCore = 0; // by default, use the first core of cpuList.
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: targetCore ?

Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: OK.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 3]
```diff
@@ -2,84 +2,185 @@
 
 import com.intel.analytics.bigdl.mkl.hardware.platform.IAAffinity;
 
-import java.util.Arrays;
-import java.util.BitSet;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class LinuxAffinity implements IAAffinity {
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: why `IAAffinity` not `IAffinity` ?

Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: IA means Intel Architecture.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 4]
```diff
@@ -2,84 +2,185 @@
...
-                int[] set = {i};
-                setAffinity0(set);
-                availableCPUs[i] = true;
-                break;
+    public synchronized void setAffinity() {
+        int min = Integer.MAX_VALUE;
+        int targetCore = 0; // by default, use the first core of cpuList.
+        for (int key : this.cpuList) {
+            int size = availableCores.get(key).size();
+            if (size < min) {
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: so here you want to select core which has least binding threads ? why not use `TreeHashMap` then

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 5]
```diff
@@ -2,84 +2,185 @@
...
+    private ConcurrentHashMap<Integer, List<Long>> availableCores;
 
     public LinuxAffinity(int cpuCounts) {
         this.cpuCounts = cpuCounts;
-        this.availableCPUs = new boolean[cpuCounts];
-        Arrays.fill(this.availableCPUs, false);
+
+        this.cpuList = getAffinity();
+
+        this.availableCores = new ConcurrentHashMap<Integer, List<Long>>();
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: what's the difference between cpu and core in your code context ?

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java (hunk 4 above)
Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: Yes. Do you mean `TreeMap`? It's not thread safe.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java (hunk 5 above)
Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: The `cpuList` term comes from affinity. I have changed them to cores here.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 6]
```diff
@@ -2,84 +2,185 @@
...
+    public synchronized void setAffinity(int[] set) {
+        boolean isAllAvailable = true;
+        for (int key : set) {
+            if (!availableCores.containsKey(key)) {
+                isAllAvailable = false;
+            }
+        }
+
+        if (!isAllAvailable) {
+            throw new BindException(set);
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: does this mean that one core can only be binded once ?

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java [hunk 7]
```diff
@@ -2,84 +2,185 @@
...
+
+        for (int i = 0; i < set.length; i++) {
+            if (set[i] == 1) {
+                temp.add(i);
             }
         }
 
-        return counts;
+        int[] result = new int[temp.size()];
+        for (int i = 0; i < temp.size(); i++) {
```
Comment (wzhongyuan) [inline native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: just `temp.toArray()` ? and then rename temp to `result`

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java (hunk 6 above)
Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: No. What does it do is checking the validation of `set`.

File: native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java (hunk 7 above)
Comment (i8run) [reply native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java]: `toArray` will return an `Object` array, which is not the same as `[]`.

**Pull request merged**